solr/solrj/src/test/org/apache/solr/client/solrj/SolrExampleTests.java

```diff
@@ -571,6 +572,57 @@ public void testGetEmptyResults() throws Exception {
     assertEquals(0, out.get(1).size());
   }
 
+  @Test
+  public void testMatchAllPaging() throws Exception {
+    SolrClient client = getSolrClient();
+
+    // Empty the database...
+    client.deleteByQuery("*:*"); // delete everything!
+    if (random().nextBoolean()) {
+      client.commit();
+    }
+    // Add eleven docs
+    List<SolrInputDocument> docs = new ArrayList<>();
+    for (int i = 0; i < 11; i++) {
+      SolrInputDocument doc = new SolrInputDocument();
+      doc.addField("id", "id" + i);
+      doc.addField("name", "doc" + i);
+      doc.addField("price", "" + i);
+      docs.add(doc);
+      if (rarely()) {
+        client.commit();
+      }
+    }
+    client.add(docs);
+    if (random().nextBoolean()) {
+      client.commit();
+    } else {
+      client.optimize();
+    }
+    final List<String> sorts = Arrays.asList("_docid_", "id", "name", "price", null);
+    Collections.shuffle(sorts, random());
+    final List<Integer> starts = Arrays.asList(0, 1, 2, 10, 11, 12);
+    Collections.shuffle(starts, random());
+    for (String sort : sorts.subList(0, 1 + random().nextInt(sorts.size() - 1))) {
+      for (int start : starts.subList(0, 1 + random().nextInt(starts.size() - 1))) {
+        final SolrQuery query = new SolrQuery("*:*");
+        if (sort != null) {
+          query.setSort(sort, random().nextBoolean() ? SolrQuery.ORDER.asc : SolrQuery.ORDER.desc);
+        }
+        if (start > 0 || random().nextBoolean()) {
+          query.setStart(start);
+        }
+        SolrDocumentList results = client.query(query).getResults();
+        assertEquals(11, results.getNumFound());
+        assertEquals("page from " + start, Math.max(Math.min(10, 11 - start), 0), results.size());
```

Comment:
Hmm, not sure about the `Math.max(Math.min(10, 11 - start), 0)` expression e.g. for `start=0` it would evaluate to `10` but paging from `0` would return `11` results, no?

Reply:
but, default `rows=0`, it will cut 11 results to 10-items page.

Reply:
ah, hadn't spotted about implicit `rows=` usage there, thanks!

how about explaining the connection like this?
```suggestion
        assertEquals("page from " + start, Math.max(Math.min(CommonParams.ROWS_DEFAULT, 11 - start), 0), results.size());
```